[ty] Adjust scope completions to use all reachable symbols

BurntSushi opened the pull request:
This PR adds a new `UseDefMap::all_reachable_symbols_with_definitions`
API for use by the LSP for offerring scoped completions. In particular,
this API uses all reachable symbols instead of only those that are
reachable at the end of the scope. Notably, this fixes a problem where
completions for local variables in functions with a `return` statement
are no longer offered.

(I opened this as a draft because I'm uncertain of the approach, and
because this needs more tests.)

Note that the first commit is just a rename. So reviewers might want to
look at this PR commit-by-commit.

Fixes https://github.com/astral-sh/ty/issues/1294

carljm commented on crates/ty_python_semantic/src/semantic_index/use_def.rs:
```diff
@@ -590,6 +590,46 @@ impl<'db> UseDefMap<'db> {
+    ) -> impl Iterator<Item = (ScopedSymbolId, Type<'db>, Definition<'db>)> {
+        self.reachable_definitions_by_symbol
+            .iter_enumerated()
+            .flat_map(|(symbol_id, reachable_definition)| {
```
nit: the second element here is a `&ReachableDefinitions`, which stores multiple bindings and declarations -- so it seems confusing to name the variable `reachable_definition` (singular)
```suggestion
            .flat_map(|(symbol_id, reachable_definitions)| {
```

BurntSushi commented:
OK, I've taken another stab at this, incorporating your feedback. Thank you. :-) Some notes though:

First is that I've left the existing `all_members_of_scope` (renamed to `all_end_of_scope_members`) in place, since it looks like places other than completions use it.

Second is that I don't think I've accounted for [this](https://github.com/astral-sh/ty/issues/1294#issuecomment-3628426664):

> Switching from "end-of-scope" to "all-reachable" for module-global scopes would probably not be desirable

Can you say more about how I should tackle this? And in particular, why is module-global scope special here? I've added a test case matching your example, but I also added a test case in function scope and it is similarly wonky. The end result is that we end up with imprecise types. Which is maybe what you were referring to on Discord and probably something we should just live with for now?

My last thought is that I'm wondering if there is an inconsistency in the terminology used for the use-def docs and its routines. In particular, reading this:

```rust
//! Let's take this code sample:
//!
//! ```python
//! x = 1
//! x = 2
//! y = x
//! if flag:
//!     x = 3
//! else:
//!     x = 4
//! z = x
//! ```
//!
//! In this snippet, we have four bindings of `x` (the statements assigning `1`, `2`, `3`, and `4`
//! to it), and two uses of `x` (the `y = x` and `z = x` assignments). The first binding of `x`
//! does not reach any use, because it's immediately replaced by the second binding, before any use
//! happens. (A linter could thus flag the statement `x = 1` as likely superfluous.)
```

That is, this verbiage is using the word "reach" here to, I believe, imply that the first binding of `x` is _not reachable_ at its first use in `y = x`. Similarly, I would say that:

```
zqzqzq: int = 1
zqzqzq: str = 'foo'
zqzq<CURSOR>
```

The first binding of `zqzqzq` is _not reachable_ on the line containing the cursor. However, the use-def map's `reachable_definitions_by_symbol` map, notably including "reachable," seems to specifically contain both declarations of `zqzqzq` in the above example.

So I guess my question is: is my understanding wrong or is the terminology not quite consistent?

carljm commented:
The new `all_reachable_symbols` API and the renaming of existing API (and leaving `all_end_of_scope_members`) all look great to me!

BurntSushi commented:
Sweet, okay. I think that all mostly makes sense to me. I've created a [follow-up issue for eager nested scopes](https://github.com/astral-sh/ty/issues/1849). Thank you for explaining all of that. I've pulled this out of draft. :-)
